### PR TITLE
Metadata API: Avoid raising securesystemslib exceptions

### LINKED ISSUE #1646: metadata API: avoid raising securesystemslib exceptions
I think Metadata API should not require its users to import securesystemlib exceptions to be able to call Metadata API methods, it just seems weird -- as a user of TUF I am not necessarily directly using securesystemslib at all, yet I need to handle exceptions from it

**Current behavior**:

sign(), from_file() and to_file() are documented to raise these exceptions

**Expected behavior**:

I should only need to import modules from tuf to be able to handle tuf exceptions. This could mean either 
* re-exporting the securesystemslib exceptions in tuf metadata or exceptions module or
* handling the securesystemslib exceptions and raising something in python-tuf: this is what I think `verify_delegate()` does

### PR DESCRIPTION
**Pull request by MVrachev**
Fixes #1646 

**Description of the changes being introduced by the pull request**:

In this pr, I aim to remove the need of our users to import securesystemslib
in order to catch the exceptions thrown by `tuf/metadata/api.py`. 
I either reexport securesystemslib exceptions or
reraise them by using tuf exceptions.

Reexporting securesystemslib exceptions is required so that users
of code that throws these exceptions doesn't have to import
securesystemslib to handle them.

A couple of additional fixes are added to this pr as well.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: I mentioned this elsewhere already but after looking at this for a while I've convinced myself that we could just handle both of them and return something generic like UnsignedMetadataError like the verify functions do, no? then I think we we can remove UnsupportedAlgorithmError altogether.

Comment (jku) [reply tuf/api/metadata.py]: ah no, it is still used by TargetFile.from_data()

**Comment by jku** [comment]
I think we can remove the last SSLIB error as well https://github.com/theupdateframework/python-tuf/pull/1725#discussion_r780236043

opinions?

**Comment by joshuagl** [comment]
> I think we can remove the last SSLIB error as well [#1725 (comment)](https://github.com/theupdateframework/python-tuf/pull/1725#discussion_r780236043)
> 
> opinions?

:+1:
Removing all sslib exceptions from the python-tuf public API seems like a good goal.

**Comment by MVrachev** [comment]
@jku and @lukpueh you can have a look into my list commit where I actually get rid of all securesystemslibs exceptions and there is no longer need for reexporting.

**Comment by lukpueh** [comment]
> @jku and @lukpueh you can have a look into my list commit where I actually get rid of all securesystemslibs exceptions and there is no longer need for reexporting.

Is this ready for review? Or should I wait until we land #1725?

**Comment by MVrachev** [comment]
You can review the last three commits.
Even though https://github.com/theupdateframework/python-tuf/pull/1734/commits/8403a891b8749fde4a7415b440ae7029f7dd6b3d and https://github.com/theupdateframework/python-tuf/pull/1734/commits/9090da2a797509624756e27b4ef6b5681297e3a1 should be squashed.

**Comment by MVrachev** [comment]
@jku @lukpueh this pr is ready for review after the merge of #1725.

File: tests/test_api.py
Comment (jku) [inline tests/test_api.py]: this comment is incorrect (sslib does not raise UnsignedMetadataError) and very long

**Comment by jku** [comment]
I think there's a misunderstanding. My comments on securesystemslib.StorageError vs OSError were made in the context of the client: there a securesystemslib error just makes life harder for no good reason. In Metadata API there was a conscious decision to abstract storage (for the repository use case). Removing StorageError completely makes no sense here: we might want to expose StorageError in our exceptions though so that users don't need to import SSLIB explicitly.

I agree about the other SSLIB errors.

File: tests/test_api.py
Comment (MVrachev) [reply tests/test_api.py]: Ohhh... such a bad comment. :D 
Yea, I am definitely updating it.

**Comment by MVrachev** [comment]
@jku the pr is updated.
You are correct we do want to keep the `StorageError` exceptions as we do provide an option to use the `storage_backend` argument and we want to be able to throw exceptions a `StorageError` which generalizes exceptions from different storage interfaces.

**Review by jku — APPROVED** [review]
I agree with this. Would appreciate other maintainer opinions (@lukpueh ?) on my suggestion to re-export StorageError: I feel like that's a good strategy but maybe a little unusual?

File: tuf/api/metadata.py
Comment (lukpueh) [inline tuf/api/metadata.py:306]: This is not raised anymore. Can you move the serialize call out of the try block?

File: tuf/api/metadata.py
Comment (lukpueh) [inline tuf/api/metadata.py]: This `except` block assumes that `signer` is an `SSlibSigner`, but the caller could pass any implementation of the `Signer`  interface. Ideally the `Signer` interface would define what exceptions to expect here, like storage or serialization interfaces do (maybe this is worth a ticket in securesystemslib?).

Should we just except all and re-raise as UnsignedMetadataError?

**Review by lukpueh — COMMENTED** [review]
Regarding `StorageError`, I think re-exporting makes sense. See inline for more comments...

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py]: good catch. For this PR (and 1.0) the last suggestion sounds reasonable to me.

File: tuf/api/metadata.py
Comment (lukpueh) [reply tuf/api/metadata.py:306]: My comment from above is nonsense. What I wanted to say is, if we except all below when calling `sign(serialize...)` and re-raise as `UnsignedMetadataError`, then `SerializationError` won't be raised anymore.

So either we take the call to serialize out of the try block below or we update the docstring

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: Really great catch.
Totatlly forgot that `signed_serializer` is an object implementing the general `securesystemslib.signer.Signer` interface.

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py:306]: I decided to move the serialization outside the try block so that the serialization error will propagate as I don't think it's good to hide it with `UnsignedMetadataError` exception.

Comment (jku) [reply tuf/api/metadata.py:306]: wow I completely missed this. Good solution I think.

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: > Ideally the Signer interface would define what exceptions to expect here, like storage or serialization interfaces do (maybe this is worth a ticket in securesystemslib?).


The `securesystemslib` issue about this: https://github.com/secure-systems-lab/securesystemslib/issues/386

**Review by lukpueh — APPROVED** [review]

**Pull request merged**
